src/WebJobs.Script/Workers/Http/HttpWorkerFunctionProvider.cs

```diff
@@ -0,0 +1,184 @@
+            _hostFunctionMetadataProvider = hostFunctionMetadataProvider ?? throw new ArgumentNullException(nameof(hostFunctionMetadataProvider));
+            _languageWorkerOptions = languageWorkerOptions ?? throw new ArgumentNullException(nameof(languageWorkerOptions));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        public ImmutableDictionary<string, ImmutableArray<string>> FunctionErrors =>
+           _errors.ToImmutableDictionary(k => k.Key, v => v.Value.ToImmutableArray());
+
+        public async Task<ImmutableArray<FunctionMetadata>> GetFunctionMetadataAsync()
+        {
+            if (!string.Equals(_httpWorkerOptions.WorkerRuntime, ScriptConstants.CustomHandlerWorkerRuntime, StringComparison.OrdinalIgnoreCase))
+            {
+                return [];
+            }
+
+            var routes = _httpWorkerOptions.HttpRoutes;
+            if (routes is null || !routes.Any())
+            {
+                return [];
+            }
+
+            var hostFunctionMetadata = await _hostFunctionMetadataProvider.GetFunctionMetadataAsync(_languageWorkerOptions.CurrentValue.WorkerConfigs, forceRefresh: false);
+
+            // We already know custom handler http routes are configured, if function.json files are also present we cannot proceed.
+            if (hostFunctionMetadata.Any())
+            {
+                throw new InvalidOperationException(
+                    "Detected both function.json files and custom handler HTTP route configuration definition in host.json" +
+                    "Only one configuration source is supported. Remove either the function.json files or the HTTP routes entries in host.json.");
+            }
+
+            return CreateFunctionsFromRoutes(routes);
+        }
+
+        private ImmutableArray<FunctionMetadata> CreateFunctionsFromRoutes(IEnumerable<HttpWorkerRoute> routes)
+        {
+            var functions = new Collection<FunctionMetadata>();
+            int i = 0;
+            foreach (var route in routes)
+            {
+                var functionName = $"http-handler{i + 1}";
+                i++;
+
+                if (!TryValidateHttpRoute(route?.Route, out string error))
+                {
+                    AddFunctionError(functionName, error);
+                    _logger.LogError("Unable to create function '{functionName}' for route '{route}' due to invalid route: {reason}", functionName, route?.Route ?? "<null>", error);
+                    continue;
+                }
+
+                functions.Add(CreateHttpFunctionMetadata(route, functionName));
+                _logger.LogInformation("Created function '{functionName}' for route '{routeTemplate}' (authLevel={auth}).", functionName, route.Route, route.AuthorizationLevel);
+            }
+
+            return [.. functions];
+        }
+
+        private static bool TryValidateHttpRoute(string route, out string error)
+        {
+            error = route switch
```

Comment:
We should be able to rely on the `TemplateParser` class for validation here. Let's not perform our own validation.